homeassistant/components/media_player/firetv.py

```diff
@@ -195,60 +155,134 @@ def source_list(self):
-            'standby': STATE_STANDBY,
-            'disconnected': STATE_UNKNOWN,
-        }.get(self._firetv.state, STATE_UNKNOWN)
-
-        if self._state not in [STATE_OFF, STATE_UNKNOWN]:
-            self._running_apps = self._firetv.running_apps
-            self._current_app = self._firetv.current_app
-        else:
-            self._running_apps = None
-            self._current_app = None
-
+        try:
+            # Check if device is disconnected.
+            if not self._firetv._adb:
+                self._state = STATE_UNKNOWN
+                self._running_apps = None
+                self._current_app = None
+
+                # Try to connect
+                self._firetv.connect()
+
+            # Check if device is off.
+            elif not self._firetv._screen_on:
+                self._state = STATE_OFF
+                self._running_apps = None
+                self._current_app = None
+
+            # Check if screen saver is on.
+            elif not self._firetv._awake:
+                self._state = STATE_IDLE
+                self._running_apps = None
+                self._current_app = None
+
+            else:
+                # Get the running apps.
+                if self._get_sources:
+                    self._running_apps = self._firetv.running_apps()
+
+                # Get the current app.
+                if self._get_source:
+                    current_app = self._firetv.current_app
+                    if isinstance(current_app, dict)\
+                            and 'package' in current_app:
+                        self._current_app = current_app['package']
+                    else:
+                        self._current_app = current_app
+
+                    # Show the current app as the only running app.
+                    if not self._get_sources:
+                        if self._current_app:
+                            self._running_apps = [self._current_app]
+                        else:
+                            self._running_apps = None
+
+                    # Check if the launcher is active.
+                    if self._current_app in [PACKAGE_LAUNCHER,
+                                             PACKAGE_SETTINGS]:
+                        self._state = STATE_STANDBY
+
+                    # Check for a wake lock (device is playing).
+                    elif self._firetv._wake_lock:
+                        self._state = STATE_PLAYING
+
+                    # Otherwise, device is paused.
+                    else:
+                        self._state = STATE_PAUSED
+
+                # Don't get the current app.
+                elif self._firetv._wake_lock:
+                    # Check for a wake lock (device is playing).
+                    self._state = STATE_PLAYING
+                else:
+                    # Assume the devices is on standby.
+                    self._state = STATE_STANDBY
+
+        except:  # pylint: disable=bare-except
```

Comment:
It is better to handle just the exceptions you are expecting to get, this will greatly help when troubleshooting.

Reply:
The reason for this `try`/`except` block and the `adb_wrapper` as a whole is because if you send too many commands in short succession, you'll get an error and the ADB connection gets broken. I'd have to experiment to figure out the different exceptions that this can lead to.

Reply:
I found 6 exceptions that occur, and now I am only catching those particular ones.